File: src/main/java/pcb/uwu/amounts/composite/finance/InterestRate.java
package pcb.uwu.amounts.composite.finance;

import pcb.uwu.core.AbstractUnitAmount;
import pcb.uwu.core.BigDecimalAmount;
import pcb.uwu.core.Magnitude;
import pcb.uwu.core.UnitAmount;
import pcb.uwu.units.composite.finance.InterestRateUnit;

import java.math.BigDecimal;
import java.math.MathContext;

public class InterestRate extends AbstractUnitAmount<InterestRateUnit> {

	// region constructors

	public InterestRate(Number value, InterestRateUnit unit) {
		super(value, unit);
	}

	public InterestRate(Number value, Magnitude magnitude, InterestRateUnit unit) {
		super(value, magnitude, unit);
	}

	public InterestRate(String value, InterestRateUnit unit) {
		super(value, unit);
	}

	public InterestRate(String value, Magnitude magnitude, InterestRateUnit unit) {
		super(value, magnitude, unit);
	}

	public InterestRate(BigDecimal value, InterestRateUnit unit) {
		super(value, unit);
	}

	public InterestRate(BigDecimal value, Magnitude magnitude, InterestRateUnit unit) {
		super(value, magnitude, unit);
	}

	public InterestRate(BigDecimalAmount amount, InterestRateUnit unit) {
		super(amount, unit);
	}

	public InterestRate(BigDecimalAmount amount, Magnitude magnitude, InterestRateUnit unit) {
		super(amount, magnitude, unit);
	}


	// endregion

	// region implement UnitAmount

	@Override
	public InterestRate plus(UnitAmount<InterestRateUnit> other, MathContext mathContext) {
		return new InterestRate(plusAmount(other, mathContext), getUnit());
	}

	@Override
	public InterestRate minus(UnitAmount<InterestRateUnit> other, MathContext mathContext) {
		return new InterestRate(minusAmount(other, mathContext), getUnit());
	}

	@Override
	public InterestRate multipliedBy(BigDecimal other, MathContext mathContext) {
		return new InterestRate(multipliedByAmount(other, mathContext), getUnit());
	}

	@Override
	public InterestRate dividedBy(BigDecimal other, MathContext mathContext) {
		return new InterestRate(dividedByAmount(other, mathContext), getUnit());
	}

	@Override
	public InterestRate convertTo(InterestRateUnit unit) {
		return new InterestRate(getAmountIn(unit), unit);
	}

	// endregion
}

File: dimension/src/main/java/pcb/uwu/amount/derived/economics/InterestRate.kt
package pcb.uwu.amount.derived.economics

import pcb.uwu.core.CompositeUnitAmount
import pcb.uwu.core.Magnitude
import pcb.uwu.core.Magnitude.NATURAL
import pcb.uwu.core.UnitAmount
import pcb.uwu.unit.derived.acoustics.FrequencyUnit
import pcb.uwu.unit.derived.economics.CurrencyUnit
import pcb.uwu.unit.derived.economics.RentUnit
import pcb.uwu.util.UnitAmountUtils

class InterestRate : CompositeUnitAmount<FrequencyUnit>
{
    @JvmOverloads
    constructor(amount: Number,
                magnitude: Magnitude = NATURAL,
                unit: FrequencyUnit)
            : super(amount, magnitude, unit)

    @JvmOverloads
    constructor(amount: String,
                magnitude: Magnitude = NATURAL,
                unit: FrequencyUnit)
            : super(amount, magnitude, unit)

    // region UnitAmount

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun plus(interestRate: UnitAmount<FrequencyUnit>) =
        InterestRate(amount = this.amount + (interestRate to this.unit).amount,
                     unit = this.unit)

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun minus(interestRate: UnitAmount<FrequencyUnit>) =
        InterestRate(amount = this.amount - (interestRate to this.unit).amount,
                     unit = this.unit)

    override fun times(scalar: Number) =
        InterestRate(amount = this.amount * scalar,
                     unit = this.unit)

    override fun div(scalar: Number) =
        InterestRate(amount = this.amount / scalar,
                     unit = this.unit)

    override fun to(unit: FrequencyUnit) =
        InterestRate(amount = UnitAmountUtils.getAmountIn(unitAmount = this, newUnit = unit),
                     unit = unit)

    // endregion

    // region composition

    fun times(currency: Currency) =
        Rent(amount = UnitAmountUtils.times(this, currency.amount.amount),
             unit = RentUnit(interestRateUnit = this.unit,
                             currency = currency.unit))

    fun times(debt: Debt) =
        Currency(amount = (this * debt).amount,
                 unit = this.unit.unitCounter.findUnit(CurrencyUnit::class)!!)

    // endregion
}
